pandas/_libs/lib.pyx

```diff
@@ -1732,6 +1740,31 @@ def is_datetime_with_singletz_array(values: ndarray) -> bool:
     return True
 
 
+def is_datetime_with_tz_array(values: ndarray) -> bool:
```

Comment:
Could you create a `DatetimeTZValidator` in the same vein as `DatetimeValidator` in the same file instead?

Reply:
I will look into that, also in the example below the infered dtype is "datetime64" although it should be "datetimetz" I think.

```python
s = pd.Series(
    [
        pd.Timestamp("2013-01-01 13:00:00-0800", tz="US/Pacific"),
        pd.Timestamp("2013-01-02 14:00:00-0800", tz="US/Pacific"),
    ]
)
assert s.dtype == "datetime64[ns, US/Pacific]"
assert infer_dtype(s, skipna=True) == "datetime64"
```

Reply:
note that we actually might want 2 different returns here (this is a general statement)

1) datetimetz_same - all the entries are the same tz
2) datetimetz_mixed - all entries have a tz but can be different - note that mixing a date time with no time zone would not be valid here (and ultimately would be mixed)

Reply:
@mroeschke I made a DatetimeTZValidator (according to my tests -see code block after the one below- it works).

```
cdef class DatetimeTZValidator(TemporalValidator):
    cdef bint is_value_typed(self, object value) except -1:
        return PyDateTime_Check(value) and (value.tzinfo is not None)

    cdef inline bint is_valid_null(self, object value) except -1:
        return is_null_datetime64(value)
```

There is still one point to clear concerning infer_dtype: as much as I like the idea from @jreback I think the distinction between datetime64 and datetime is more important right now and that we need to add a "datetime64tz" return value for inder_dtype. Later on we could then implement "datetimetz_same", "datetimetz_mixed", "datetimet64tz_same", "datetime64tz_mixed".

I prepared some tests and checked if I was able to modify pandas._libs.lib to achieve the distinctions "datetimetz"/"datetime64tz"/"datetime"/"datetime64" and actually got it to work. There is definitely room for improvements and I will make a commit later (unless you do not like this idea). Anyhow here are the tests I checked:

```python
from pandas.api.types import infer_dtype
import pandas as pd
import numpy as np
import datetime

assert infer_dtype([datetime.datetime(2019,1,1,0,0,0, 
                                      tzinfo = datetime.timezone.utc)]) == 'datetimetz'
assert infer_dtype([pd.Timestamp('2019-01-01', tz = 'UTC')]) == 'datetimetz'
assert infer_dtype([pd.Timestamp('2019-01-01', 
                                 tzinfo = datetime.timezone.utc)]) == 'datetimetz'

# mixed (with and without tz will return datetime)
assert infer_dtype([pd.Timestamp('2019-01-01', 
                                 tzinfo = datetime.timezone.utc),
                    pd.Timestamp('2019-01-01')]) == 'datetime'
assert infer_dtype([datetime.datetime(2019,1,1,0,0,0)]) == 'datetime'
assert infer_dtype([pd.Timestamp('2019-01-01')]) == 'datetime'

# this monstrocity will return "datetimetz"
assert infer_dtype([pd.Timestamp('2019-01-01', 
                                 tzinfo = datetime.timezone.utc),
                    datetime.datetime(2019,1,1,0,0,0, 
                                      tzinfo = datetime.timezone.utc)]) == "datetimetz"

s = pd.Series([pd.Timestamp("2013-01-01 13:00:00-0800", 
                            tz="US/Pacific"),
               pd.Timestamp("2013-01-02 14:00:00-0800", 
                            tz="US/Pacific")])
assert infer_dtype(s, skipna=True) == "datetime64tz"

s = pd.Series([pd.Timestamp("2013-01-01"),
               pd.Timestamp("2013-01-02")])
assert infer_dtype(s, skipna=True) == "datetime64"

arr = np.array([datetime.datetime(2019,1,1,0,0,0, 
                                  tzinfo = datetime.timezone.utc),
                datetime.datetime(2019,1,1,0,0,0, 
                                  tzinfo = datetime.timezone.utc)])
assert infer_dtype(arr, skipna=True) == "datetime64tz"
```

Also I wanted to say sorry for the mess I still need to learn this is my first PR and first time really working with Cython. I am focusing on infer_dtype before fixing the errors in the Azure pipelines (I already observed them though).

Reply:
don’t distinguish datetime64 ; these are just things that are convertible to date time dtypes

make your changes S simple as possible